From 7f8e86a364d7a2b5ec8bbae87b40f4411fe8fac1 Mon Sep 17 00:00:00 2001
From: Yu Chien Peter Lin <peterlin@andestech.com>
Date: Fri, 15 Dec 2023 17:19:00 +0800
Subject: [PATCH] rzfive: enable caches

1. The mcache_ctl.CCTL_SUEN should be set so
   ax45mp_iocp_sw_workaround() can return true.
2. Enable L2-cache, along with some settings

bit#:   32109876 5 4 3 21 0 98 7 65 43 21 0
0x900   00000000 0 0 0 01 0 01 0 00 00 00 0  Gen1 reset
0x979   00000000 0 0 0 01 0 01 0 11 11 00 1  Gen1 boot up
                   | |  | |  | |  |  |  | `- (RW) CEN    : L2-Cache enbale
                   | |  | |  | |  |  |  `--- (RW) PFTHRES: Prefetch threshold
                   | |  | |  | |  |  `------ (RW) IPFDPT : Instruction prefetch depth
                   | |  | |  | |  `--------- (RW) DPFDPT : Data prefetch depth
                   | |  | |  | `------------ (RW) ECCEN  : ECC enable
                   | |  | |  `-------------- (RW,RO) TRAMOCTL:  Tag RAM output cycle
                   | |  | `----------------- (RW,RO) TRAMICTL:  Tag RAM setup  cycle
                   | |  `------------------- (RW,RO) DRAMOCTL: Data RAM output cycle
                   | `---------------------- (RW,RO) DRAMICTL: Data RAM setup  cycle
                   `------------------------ (RO) INITSTATUS : Self-initialization status

These cache oprations might be moved to U-Boot SPL,
we will update the patch once U-Boot patches are
posted on ML.

Upstream-Status: Inappropriate [enable feature]
Signed-off-by: Yu Chien Peter Lin <peterlin@andestech.com>
---
 platform/generic/renesas/rzfive/rzfive.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/platform/generic/renesas/rzfive/rzfive.c b/platform/generic/renesas/rzfive/rzfive.c
index b5b5e65..4c7bfaf 100644
--- a/platform/generic/renesas/rzfive/rzfive.c
+++ b/platform/generic/renesas/rzfive/rzfive.c
@@ -7,6 +7,8 @@
 #include <andes/andes45_pma.h>
 #include <andes/andes_pmu.h>
 #include <andes/andes_sbi.h>
+#include <andes/andes45.h>
+#include <sbi/riscv_asm.h>
 #include <platform_override.h>
 #include <sbi/sbi_domain.h>
 #include <sbi_utils/fdt/fdt_helper.h>
@@ -32,6 +34,9 @@ static int renesas_rzfive_final_init(bool cold_boot, const struct fdt_match *mat
 
 static int renesas_rzfive_early_init(bool cold_boot, const struct fdt_match *match)
 {
+#define L2C_CTR 0x13400008
+	csr_write(CSR_MCACHE_CTL, 0x183f03ul);
+	*(unsigned long *)L2C_CTR = 0x979;
 	/*
 	 * Renesas RZ/Five RISC-V SoC has Instruction local memory and
 	 * Data local memory (ILM & DLM) mapped between region 0x30000
-- 
2.34.1

